lib/core/src/restore.rs

```diff
@@ -0,0 +1,820 @@
+//! This module provides functionality for restoring the swap tx IDs from onchain data
+
+use std::collections::HashMap;
+
+use anyhow::Result;
+use log::{error, info};
+use lwk_wollet::elements::Txid;
+use lwk_wollet::WalletTx;
+
+use crate::restore::immutable::*;
+use crate::sdk::LiquidSdk;
+
+/// A map of all our known LWK onchain txs, indexed by tx ID
+pub(crate) struct TxMap {
+    outgoing_tx_map: HashMap<Txid, WalletTx>,
+    incoming_tx_map: HashMap<Txid, WalletTx>,
+}
+impl TxMap {
+    pub(crate) fn from_raw_tx_map(raw_tx_map: HashMap<Txid, WalletTx>) -> Self {
+        let (outgoing_tx_map, incoming_tx_map): (HashMap<Txid, WalletTx>, HashMap<Txid, WalletTx>) =
+            raw_tx_map
+                .into_iter()
+                .partition(|(_, tx)| tx.balance.values().sum::<i64>() < 0);
+
+        Self {
+            outgoing_tx_map,
+            incoming_tx_map,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainDataSend {
+    lockup_tx_ids: HashMap<String, Txid>,
+    refund_tx_ids: HashMap<String, Txid>,
+}
+
+pub(crate) struct RecoveredOnchainDataReceive {
+    lockup_claim_tx_ids: HashMap<String, (Txid, Txid)>,
+}
+
+pub(crate) struct RecoveredOnchainDataChainSend {
+    lbtc_user_lockup_tx_ids: HashMap<String, Txid>,
+    lbtc_refund_tx_ids: HashMap<String, Txid>,
+    btc_server_lockup_tx_ids: HashMap<String, Txid>,
+    btc_claim_tx_ids: HashMap<String, Txid>,
+}
+
+pub(crate) struct RecoveredOnchainDataChainReceive {
+    /// Server lockup tx ID, claim tx ID.
+    ///
+    /// We store them in a pair because when they are present, we always expect both to be present.
+    lbtc_server_lockup_claim_tx_ids: HashMap<String, (Txid, Txid)>,
+
+    btc_user_lockup_tx_ids: HashMap<String, Txid>,
+    btc_refund_tx_ids: HashMap<String, Txid>,
+}
+
+pub(crate) struct RecoveredOnchainData {
+    send: RecoveredOnchainDataSend,
+    receive: RecoveredOnchainDataReceive,
+    chain_send: RecoveredOnchainDataChainSend,
+    chain_receive: RecoveredOnchainDataChainReceive,
+}
+
+impl LiquidSdk {
+    pub(crate) async fn recover_from_onchain(&self, tx_map: TxMap) -> Result<()> {
+        let immutable_db = self.get_swaps_list().await?;
```

Comment:
NIT: instead of immutable_db perhaps just `swaps_scripts_list`? or `swaps_list`?

Reply:
Also it is better to call this function out of the recover_from_onchain since I can see later that we will need to invoke it on a list that comes from realtime sync (and doesn't exist in the db).

Reply:
Done: da39009268bde34d26b81a505873e8f028acaf0b